pkg/frontend/v2/frontend_test.go

```diff
@@ -332,8 +339,82 @@ func TestFrontendFailedCancellation(t *testing.T) {
 	require.Nil(t, resp)
 
 	ms.checkWithLock(func() {
-		require.Equal(t, 1, len(ms.msgs))
+		require.Len(t, ms.msgs, 1)
+	})
+}
+
+// Test that the front-end shuts down gracefully, i.e. waits for in-flight queries to finish.
+func TestFrontend_GracefulShutdown(t *testing.T) {
+	const (
+		userID = "test"
+		body   = "all good"
+	)
+	ctx := user.InjectOrgID(context.Background(), userID)
+
+	setupDone := make(chan uint64)
+	t.Cleanup(func() {
+		close(setupDone)
 	})
+	f, _ := setupFrontend(t, nil, func(f *Frontend, msg *schedulerpb.FrontendToScheduler) *schedulerpb.SchedulerToFrontend {
+		go func() {
+			// Allow for the service to register a successful enqueuing
+			time.Sleep(100 * time.Millisecond)
+			setupDone <- msg.QueryID
+		}()
+
+		return &schedulerpb.SchedulerToFrontend{Status: schedulerpb.OK}
+	})
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		t.Helper()
+
+		t.Log("sending request")
+		resp, err := f.RoundTripGRPC(ctx, &httpgrpc.HTTPRequest{})
+		require.NoError(t, err)
+		t.Log("received response", "code", resp.Code)
+		require.Equal(t, int32(200), resp.Code)
+		require.Equal(t, []byte(body), resp.Body)
+		wg.Done()
+	}()
+
+	t.Log("waiting on service to register enqueuing")
+	queryID := <-setupDone
+	t.Log("finished waiting on enqueuing")
+
+	// Wait for the request to be handled, and make sure the service doesn't terminate until the request
+	// is no longer in flight.
+
+	var serviceStopped atomic.Bool
```

Comment:
Can we check for `f.State() == services.Terminated` instead?

Reply:
Good observation, but I guess this is irrelevant after I pushed your revised test.